internal/controller/nab_validator.go

```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+
+	"github.com/go-logr/logr"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	nacv1alpha1 "github.com/migtools/oadp-non-admin/api/v1alpha1"
+	"github.com/migtools/oadp-non-admin/internal/common/constant"
+	"github.com/migtools/oadp-non-admin/internal/common/function"
+)
+
+// ValidateVeleroBackupSpec validates the VeleroBackup Spec from the NonAdminBackup.
+//
+// Parameters:
+//
+//	ctx: Context for the request.
+//	log: Logger instance for logging messages.
+//	nab: Pointer to the NonAdminBackup object.
+//
+// The function attempts to get the BackupSpec from the NonAdminBackup object.
+// If an error occurs during this process, the function sets the NonAdminBackup status to "BackingOff"
+// and updates the corresponding condition accordingly.
+// If the BackupSpec is invalid, the function sets the NonAdminBackup condition to "InvalidBackupSpec".
+// If the BackupSpec is valid, the function sets the NonAdminBackup condition to "BackupAccepted".
+func (r *NonAdminBackupReconciler) ValidateVeleroBackupSpec(ctx context.Context, log logr.Logger, nab *nacv1alpha1.NonAdminBackup) (exitReconcile bool, requeueReconcile bool, errorReconcile error) {
+	logger := log.WithValues("NonAdminBackup", nab.Namespace)
+
+	// Main Validation point for the VeleroBackup included in NonAdminBackup spec
+	_, err := function.GetBackupSpecFromNonAdminBackup(nab)
+
+	if err != nil {
+		errMsg := "NonAdminBackup CR does not contain valid BackupSpec"
+		logger.Error(err, errMsg)
+
+		updatedStatus, errUpdateStatus := function.UpdateNonAdminPhase(ctx, r.Client, logger, nab, nacv1alpha1.NonAdminBackupPhaseBackingOff)
+		if errUpdateStatus != nil {
+			logger.Error(errUpdateStatus, "Unable to set NonAdminBackup Phase: BackingOff", nameField, nab.Name, constant.NameSpaceString, nab.Namespace)
+			return true, false, errUpdateStatus
+		} else if updatedStatus {
```

Comment:
this part is happy path, no errors happened, right :question: should not it set condition prior to exiting :question:

Reply:
No, the happy path is the code after 57 line this is semi-happy path where the status was updated, but the reconcile should continue with different path if the status did not require update.

Reply:
No. We do not re-queue on status or condition update. The 67 did modify condition already and the reconcile loop got status update so we reconciled again, meaning this is second time this function is called.

That is why we have `updatedStatus` flag to ensure we don't go into infinite loop of reconciles and only update when needed.

The status update and condition update are separate and independent this way and it's better as they may be populated in different workers / times.